core/src/main/java/org/elasticsearch/env/NodeEnvironment.java

```diff
@@ -185,8 +187,14 @@ public NodeEnvironment(Settings settings, Environment environment) throws IOExce
             IOException lastException = null;
             int maxLocalStorageNodes = MAX_LOCAL_STORAGE_NODES_SETTING.get(settings);
             for (int possibleLockId = 0; possibleLockId < maxLocalStorageNodes; possibleLockId++) {
-                for (int dirIndex = 0; dirIndex < environment.dataWithClusterFiles().length; dirIndex++) {
-                    Path dir = environment.dataWithClusterFiles()[dirIndex].resolve(NODES_FOLDER).resolve(Integer.toString(possibleLockId));
+                for (int dirIndex = 0; dirIndex < environment.dataFiles().length; dirIndex++) {
+                    Path dataDir = environment.dataFiles()[dirIndex];
+                    // Remove the cluster name from the data path if it exists
+                    if (dataPathUpgradeNeeded(dataDir)) {
+                        logger.debug("upgrading data directory to not include cluster name");
+                        upgradeDataPath(logger, dataDir, environment.dataWithClusterFiles()[dirIndex]);
+                    }
+                    Path dir = dataDir.resolve(NODES_FOLDER).resolve(Integer.toString(possibleLockId));
```

Comment:
not sure how we can fix that to be honest since we can't move while holding the locks?